src/Joblify.Search/OfferSearchIndex.cs

```diff
@@ -0,0 +1,92 @@
+﻿using AutoMapper;
+using Joblify.Core.Offers;
+using Joblify.Search.Models;
+using Microsoft.Azure.Search;
+using Microsoft.Azure.Search.Models;
+using Microsoft.Extensions.Configuration;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Joblify.Search
+{
+    public class OfferSearchIndex : SearchIndexClient, IOfferSearchIndex
+    {
+        private readonly IOfferService _offerService;
+        private readonly SearchServiceClient _searchService;
+        private readonly IMapper _mapper;
+        private readonly string _indexName;
+
+        public OfferSearchIndex(IConfiguration configuration, IMapper mapper, IOfferService offerService)
+            : base(configuration["SearchServiceName"], configuration["OfferSearchIndexName"],
+                  new SearchCredentials(configuration["SearchServiceQueryApiKey"]))
+        {
+            _searchService = new SearchServiceClient(configuration["SearchServiceName"],
+              new SearchCredentials(configuration["SearchServiceAdminApiKey"]));
+            _indexName = configuration["OfferSearchIndexName"];
+            _offerService = offerService;
+            _mapper = mapper;
+        }
+
+        private void CreateIndex()
+        {
+            var definition = new Index()
+            {
+                Name = _indexName,
+                Fields = FieldBuilder.BuildForType<OfferSearchModel>()
+            };
+
+            _searchService.Indexes.Create(definition);
+        }
+
+        private bool AddBatchToIndex(IndexBatch<OfferSearchModel> offerBatch, int count = 0)
+        {
+            try
+            {
+                Documents.Index(offerBatch);
+                return true;
+            }
+            catch (IndexBatchException)
+            {   // sometimes adding to index may fail 
+                // (may happen if service is under heavy load)
+                if (count < 5)
+                {
+                    Thread.Sleep(5000);
+                    return AddBatchToIndex(offerBatch, ++count);
+                }
+                return false;
+            }
+        }
+
+        public async Task<OfferDto> AddOfferAsync(OfferDto offerDto)
```

Comment:
I think in the end approach should be a bit different: you leave adding offer in it's service as it is rn, and after ComitAsync in offer sevice, this method should be called with result dto as parameter. This way indexing still depends on adding offer process, but is not responsible for actually performing it.